---
title: "DMM_clustering"
author: "ruben"
date: "30/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## LIbrairies

```{r,message=F,warning=F}

library(data.table)
library(devtools)
library(microbiome)
library(DirichletMultinomial)
library(reshape2)
library(magrittr)
library(phyloseq)
library(dplyr)
library(tibble)
library(rstatix)
library(ggpubr)
library(ggplot2)
library(tidyverse)
library("readxl")
library("optparse")
library(varhandle)

library(gplots)

source(file = "functions.R")

```

## Data

```{r,message=F,warning=F}

seed = 12000

set.seed(seed)

threshold = 500

# we load this file to have taxa table specifically
load(file = "study_after_QC.rda")

# we load only the phy object
load(file = "phy_curated_data_after_QC.rda")

phy <- subset_samples(phy, body_site == "stool")

## diversity vegan

phy_diversity <- estimate_richness(phy, split = TRUE, measures = "Shannon")

#phy <- subset_samples(phy, age_category == "adult")

#phy <- subset_samples(phy, disease == "Control")

phy <- subset_samples(phy, antibiotics_current_use == "no")

pseq <- subset_taxa(phy, Genus %in% c("Bifidobacterium"))


```



## Transformation

```{r,message=F,warning=F}

# phy.comp <- microbiome::transform(phy, "compositional")
# compositional transformation, ie, relative abundance

#taxa <- core_members(phy.comp, detection = 0.1/100, prevalence = 50/100)

#pseq <- prune_taxa(taxa, phy)

# Pick the OTU count matrix
# and convert it into samples x taxa format
dat <- abundances(pseq)
count_raw <- as.matrix(t(dat))

# rowsums equal zero (sample with 0 count everywhere) need to be deleted

sample_low_reads <- names(which(rowSums(count_raw) <= 500))

# colsums equal zero (bacteria not present in any sample) need to be deleted

count <- count_raw %>%
  as.data.frame(.) %>%
  filter_all(., any_vars(. != 0))

# colsums equal zero (bacteria not present in any sample) need to be deleted

otu_not_present <- names(which(colSums(count) == 0))

count <- count %>%
  select(-otu_not_present) %>%
  ## filtering ppl with nb reads < threshold
  filter(!row.names(.) %in% sample_low_reads) %>%
  as.matrix(.)

# transform colnames to species name

colnames(count) <- get_species(as.data.frame(tax),colnames(count))

```


## Code à faire marcher la nuit

```{r,message=F,warning=F}

fit_genus_list = vector("list",5)


set.seed(seed)
seeds=sample(1:1000, 5)

name <- paste0("DMM_fit_list",seed)

if(!(name %in% list.files())) {

  for(i in 1:5) {

    set.seed(seeds[i])

    fit_genus <- mclapply(1:30, dmn, count=count, verbose=FALSE, mc.cores=12)

    fit_genus_list[[i]] = fit_genus
    
    print(i)

  }
  
  save(fit_genus_list, file=name)
  
  save(seeds, "seeds_list")

} else {
  load(name)
  load("seeds_list")
  }

# collect scores to find the best fit
lplc = vector("list",5)
bic <- vector("list",5)
aic <- vector("list",5)

for(i in 1:5) {

  
  lplc[[i]] <- base::sapply(fit_genus_list[[i]], DirichletMultinomial::laplace)

  aic[[i]]  <- base::sapply(fit_genus_list[[i]], DirichletMultinomial::AIC)

  bic[[i]]  <- base::sapply(fit_genus_list[[i]], DirichletMultinomial::BIC)

}

```
## Model fit

```{r,message=F,warning=F}

## we will take the number k for each model and for each method

list_k <- vector("list",15)

for (i in 1:5) {
  
  seed_experiment <-seeds[[i]]
  
  list_k[[i]] <- c("lplc", rank(lplc[[i]]), seed_experiment)
  
  list_k[[i+5]] <- c("aic", rank(aic[[i]]), seed_experiment)
  
  list_k[[i+10]] <- c("bic", rank(bic[[i]]), seed_experiment)
  
  print(i)
  
}

test <- data.frame(t(as.data.frame(list_k)))

rownames(test) <- NULL

colnames(test) <- c("method", 1:30, "seed")


test_unfactored <- test %>%
  select('1':'30') %>%
  unfactor(.) %>%
  cbind(method = test$method, . , seed = test$seed)
rownames(test_unfactored) <- NULL
  

## calcul des rangs moyens

mean_rank <- test_unfactored %>%
  select(-seed) %>%
  group_by(method) %>%
  summarise_all(mean) %>%
  gather(., key="k", value="mean_rank", '1':'30')

```
## Plot des rangs moyens

```{r,message=F,warning=F}

dds <- phyloseq_to_deseq2(genera_temp, ~ 1)
dds_res2 <- DESeq(dds, test="Wald")
co = matrix(as.character(combn(levels(data.frame(sample_data(genera_temp))[,"DP5_FoodGroups"]),2)),nrow=2) # Generate pairs of groups
output_FDR <- matrix(0,ncol=ncol(co),nrow=length(signif_genera))
rownames(output_FDR) <- signif_genera
colnames(output_FDR) <- apply(co,MARGIN=2,paste,collapse="-")
output_LogFC <- output_FDR
for(elem in 1:ncol(co)){ # Loop on comparisons
  # Perform contrast
  res_onecomp <- results(dds_res2, contrast=c("DP5_FoodGroups",glue("{co[1,elem]}"),glue("{co[2,elem]}")) , alpha=0.05)
  # Extract unadjusted pvalue of genera with significant global effect, and adjust it with BH
  output_FDR[,elem] <- p.adjust(res_onecomp[signif_genera,"pvalue"],method="BH")
  output_LogFC[,elem] <- res_onecomp[signif_genera,"log2FoldChange"]
}

```

## Plot des rangs moyens

```{r,message=F,warning=F}

ggplot(mean_rank) +
  geom_line(aes(x=k, y=mean_rank, group=method, color =method)) +
  theme_minimal() +
  scale_x_discrete(limits = as.character(1:30))

```
## Model selection

```{r,message=F,warning=F}

## we select k=6 with BIC method

model_k <- vector("list",5)

for (i in 1:5) {
  model_k[[i]] <- which(bic[[i]] == min(bic[[i]]))
}

index <- which(as.data.frame(model_k) == 6)

## the model 1 is the bets fit model

## le fit correspond à l'indice 1

fit <- fit_genus_list[[index]]

best <- fit[[which.min(unlist(bic[[index]]))]]

## best is the best model based on BIC criteria



```


## Check figure

```{r,message=F,warning=F}

laplace_plot <- data.frame(Fit = lplc[[index]], k = c(1:length(lplc[[index]])))
aic_plot <- data.frame(Fit = aic[[index]], k = c(1:length(aic[[index]])))
bic_plot <-data.frame(Fit = bic[[index]], k = c(1:length(bic[[index]])))

## on bind les 3 par lignes

binded <- bind_rows(list(laplace = laplace_plot, aic = aic_plot, bic = bic_plot), .id = "type_check")

## the minimum for laplace check

mins_laplace <- binded %>%
  filter(type_check == 'laplace' ) %>%
  filter(Fit == min(Fit)) %>%
  select(Fit, k)

mins_bic <- binded %>%
  filter(type_check == 'bic' ) %>%
  filter(Fit == min(Fit)) %>%
  select(Fit, k)


ggplot(binded) +
  geom_line(aes(x=k, y=Fit, group=type_check, color =type_check)) +
  theme_minimal() +
  scale_x_continuous(breaks = seq(0, max(binded$k) + 5, by = 5))  +
  
  ### we add laplace minimum
  annotate("point", x=mins_laplace$k, y=mins_laplace$Fit, color='orange', size=4) +
  geom_hline(yintercept = mins_laplace$Fit, linetype='dashed') +
  annotate("label", x=max(laplace_plot$k) + 3, y=mins_laplace$Fit, color='black', label= c("laplace minimum"), size=4) +
  annotate(geom = "text", x = mins_laplace$k, y = mins_laplace$Fit, label = paste0("k = ", mins_laplace$k), vjust = -0.5) +
  
    ### we add bic minimum
  annotate("point", x=mins_bic$k, y=mins_bic$Fit, color='orange', size=4) +
  geom_hline(yintercept = mins_bic$Fit, linetype='dashed') +
  annotate("label", x=max(laplace_plot$k) + 3, y=mins_bic$Fit, color='black', label= c("bic minimum"), size=4) +
  annotate(geom = "text", x = mins_bic$k, y = mins_bic$Fit, label = paste0("k = ", mins_bic$k), vjust = -0.5) +
  
  coord_cartesian(xlim =c(min(binded$k),max(binded$k) + 6))  +
  
  labs(x = "Number of Dirichlet Component - K", y = "Model Fit", title = "", color = "")

#max(laplace_plot$k) + 10

```

## Laplace check

```{r,message=F,warning=F}

## sample component assignments

model <- apply(mixture(best), 1, which.max)

## write the enterotype file

enterotype <- data.frame(enterotype = model)

enterotype <- enterotype %>%
  rownames_to_column("sample_id")

write.csv(enterotype, file="enterotype_DMM", row.names=TRUE, col.names=TRUE, quote=FALSE)

t <- fread("enterotype_DMM") %>%
  column_to_rownames("V1")


```


## Heatmaps

```{r,message=F,warning=F,fig.width=15, fig.height=8}

matrix_int <- fitted(best) %>%
  log10(.) %>%
  round(.,2)

matrix_int <- round(matrix_int,2)

## replace col and rownames

colnames(matrix_int) <- paste("cluster ",1:ncol(matrix_int))

rownames(matrix_int) <- gsub("Bifidobacterium","B. ",rownames(matrix_int))


my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)

col_breaks = c(seq(0.81,1,length=100), # for red
seq(0,0.8,length=100),  # for yellow
seq(-1,0,length=100)) # for green

#png(filename = "heatmap_dmm.png", width = 1000, height = 1000, units = "px")

heatmap.2(matrix_int,
  cellnote = matrix_int,  # same data set for cell labels
  main = "", # heat map title
  notecol="black",      # change font color of cell labels to black
  density.info="none",  # turns off density plot inside color legend
  trace="none",         # turns off trace lines inside the heat map
  margins = c(12,9),     # widens margins around plot
  col=my_palette,       # use on color palette defined earlier
  dendrogram="none",     # only draw a row dendrogram
  Colv="NA",
  key.title = "",
  key.xlab = expression(paste("log"[10],"(Value)")))           # turn off column clustering


#dev.off()

```

## Boxplot Shannon

```{r,message=F,warning=F}

diversity <- phy_diversity %>%
  rownames_to_column("sample_id") %>%
  merge(., enterotype, by="sample_id") %>%
  select(-sample_id)

diversity$enterotype <- as.character(diversity$enterotype)

p <- ggplot(diversity, aes(x = enterotype,
                       y = Shannon)) +
  geom_boxplot(outlier.shape = NA) +
  theme_minimal() +
  stat_compare_means(label.y = 4.3, label.x ="3") +
  labs(fill= "enterotype")
  
p

```

## Description des clusters

```{r,message=F,warning=F}

distribution <- function(category, type){
  

  map <- phy@sam_data %>%
    data.frame(.) %>%
    rownames_to_column("sample_id") %>%
    merge(., enterotype, by="sample_id") %>%
    select(enterotype,all_of(category)) %>%
    group_by_all(.) %>%
    summarise(rows = n(), .groups = 'drop') %>%
    group_by_at('enterotype') %>%
    mutate(percent = rows/sum(rows)) %>%
    rename('category' = category)
  
 # map$percent <- round(map$percent*100,1)
  
  p <- ggplot(map, aes(x = enterotype,y=percent, fill = category)) +
    geom_bar(stat="identity") +
    geom_text(aes(label = paste0(round(percent*100,1),"%")),position = position_fill(vjust = 0.5), size =2.5) +
    #geom_text(aes(label = paste0(round(percent*100,1),"%"), vjust = 1), position = position_fill(vjust = 0.5), size =3) +
    theme_classic() +
    scale_fill_brewer(type = type) +
    labs(x = "Cluster K", y = "percent", fill = category)
  
  p

}

distribution('westernized', 'qual')

distribution('age_category', type = NULL)

distribution('disease', "qual")

```
## table de contingence enterotypes * bifidotypes

```{r,message=F,warning=F}

## library for chi squared test ##

library("gplots")

## enterotypes are specific to bifid, so we'll call it bifidotype, comparing it to enterotypes

bifidotype <- enterotype

enterotype_whole_microbiota <- fread('enterotypes_curated_v3_prediction.csv') %>%
  select(sample_name, Enterotypes_id) %>%
  rename(., "sample_id" = sample_name)

contingeny_table <- bifidotype %>%
  merge(., enterotype_whole_microbiota, by="sample_id") %>%
  select(- sample_id)

names(contingeny_table) <- c('bifidotypes','enterotypes')

contingeny_table <- contingeny_table %>%
  table(.)

## représentation graphique 

test_chisq <- chisq.test(contingeny_table)

test_chisq

```
## enterotypes * bifidotypes plot with pearson residuals

```{r,message=F,warning=F}
library(corrplot)

corrplot(test_chisq$residuals, is.cor = FALSE)

```

## test health_healthy

```{r,message=F,warning=F}

## we select the "control", which are "healthy" in the database :

healthy <- map %>%
  select(sample_id, disease) %>%
  filter(disease == 'Control')

print(paste0('le nombre de healthy dans curation est : ',nrow(healthy)))

healthy_healthy <- read_excel('metadata_healthy_healthy.xlsx') %>%
  rename('sample_id' = Sample_ID) %>%
  merge(., healthy, by='sample_id')
  
print(paste0('le nombre de healthy-healthy dans curation est : ',nrow(my_data)))

```

## Conseils

```{r,message=F,warning=F}


# dans le réseau, extraire le poids des liens, positifs ou négatifs

#sebeta <- symBeta(getOptBeta(MSP_spiec_easi), mode='maxabs')

## récuprer les beta les poids du réseau à partir de l'objet MSP_spiec_easi

# tableau adjacent

# https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/inference-of-microbial-ecological-networks.html




### faire un tableau du nombre de k meilleur en fonction des seeds




# 1 faire le script pour lancer la nuit


# figure de Muriel

```



