---
title: "Genes_MAGS"
author: "ruben"
date: "12/05/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Librairies

```{r,message=F,warning=F}

library(dplyr)
library(readr)
library(pheatmap)
library(data.table)
library(tidyr)
library(tibble)
library("factoextra")
library("FactoMineR")
library(RColorBrewer)
library(tidyverse)
library(broom)


source(file = "functions.R")

```

## import data

```{r,message=F,warning=F}

species <- c("Bifidobacterium adolescentis","Bifidobacterium animalis","Bifidobacterium bifidum","Bifidobacterium catenulatum","Bifidobacterium longum","Bifidobacterium pseudocatenulatum")

load("curated_v3_otu_tax.rda")

setwd("/home/tapju/storage/actibiome")

## MAGS annotations data, containing completeness, average distance according to a reference genome, etc
mags_tax_file = readr::read_tsv("data-raw/bif_mags/Bifidobacterium_annotation/annotations_url_metadata_opendata.tsv") %>%
  filter(assigned_genus=="Bifidobacterium") %>%
  filter(completeness > 80) %>%
  select(genome_name,study,sample_name,assigned_species,completeness, average_distance)




## MAGS linked to functions ID's
mags_tax_gene_id_eggnog <- readr::read_csv2("data-raw/bif_mags/mags_tax_gene_id_eggnog.csv") %>%
   select(-1) %>%
  ## filter the species of interest
   filter(grepl("adolescentis|animalis| longum|bifidum|dentium|pseudocatenulatum|catenulatum|animalis", assigned_species)) %>%
  ### eggnog is linked to several annotations KEGG, CAZy, so we link only 1 annotations for 1 eggnog
  group_by(eggNOG_OGs) %>%
  mutate(KEGG_ko = get_mode(KEGG_ko)) %>%
  mutate(CAZy = get_mode(CAZy)) %>%
  mutate(EC = get_mode(EC)) %>%
  ungroup() %>%
  filter(completeness > 80)

## number of mags per species

nb_mags_species <- mags_tax_file %>%
  data.frame(.) %>%
  group_by(assigned_species) %>%
  summarise(n=n()) %>%
  filter(grepl("adolescentis|animalis| longum|bifidum|dentium|pseudocatenulatum|catenulatum|animalis", assigned_species))
nb_mags_species$assigned_species <- gsub("Bifidobacterium","B.",nb_mags_species$assigned_species)


```

## mags _ tax file with selected species

```{r,message=F,warning=F}

sample <- sampleMetadata

## if one ppl has several samples, keep the sample with the highest nb of reads
sample <- unique(setDT(sampleMetadata)[order(subject_id, -number_reads)], by = "subject_id")

metadata <- sample %>%
   ## transform disease and westernized to good format
   transform_disease(.) %>%
   transform_westernized(.) %>%
   select(sample_id, body_site, disease, age_category, gender, country, westernized, BMI, antibiotics_current_use)


## merge with mags_tax

mags_tax <- mags_tax_file %>%
   merge(metadata, by.x = "sample_name", by.y = "sample_id") %>%
   tibble::column_to_rownames("genome_name") %>%
   filter(assigned_species %in% species)
  

```

## code block to add healthy/unhealthy clusters

```{r,message=F,warning=F}
metadata <- metadata

#setwd("/home/ladeirru/GitHub/microbiome.actibiome")
bifido <- fread("enterotype_DMM") %>%
   select(-V1) %>%
   merge(metadata, by="sample_id") %>%
   mutate(grp_healthy = case_when(
    bifidotype %in% c(1,2,6) ~ "healthy",
    TRUE ~ "unhealthy"
  ))


mags_tax <- mags_tax %>%
  rownames_to_column("MAGS") %>%
  merge(bifido %>% select(sample_id, grp_healthy), by.x = "sample_name", by.y="sample_id") %>%
  column_to_rownames("MAGS")

```

## R Markdown

```{r, message=F,warning=F, fig.height=9, fig.width = 9}

functionality <- "eggNOG_OGs"

mags_ko_file_name <- "mags_ko_file.csv"

mags_ko_table <- mags_ko_table <- readr::read_csv2(mags_ko_file_name) %>%
  merge(mags_tax_file %>% select(genome_name, assigned_species), by = "genome_name")


heat <- mags_ko_table %>%
  tibble::column_to_rownames("genome_name")
   
```


## Prepare heat df

```{r, message=F,warning=F, fig.height=9, fig.width = 9}

parameter_name <- "grp_healthy"
heat_evaluate_chi <- heat %>%
  select(-assigned_species)
heat_evaluate_chi[heat_evaluate_chi >1] <- "1"
heat_evaluate_chi[heat_evaluate_chi <1] <- "0"

#heat_evaluate_chi <- cbind(heat_evaluate_chi, assigned_species = heat$assigned_species)
#heat_evaluate_chi <- cbind(heat_evaluate_chi, assigned_species = heat$assigned_species)

```


## Are KEGG_ko, Cazy, KEGG_ko unique to eggnogs ?

```{r, message=F,warning=F, fig.height=9, fig.width = 9}

count_functions_eggnogs <- mags_tax_gene_id_eggnog %>%
  select(eggNOG_OGs, KEGG_ko) %>%
  group_by(eggNOG_OGs) %>%
  summarise(count = n_distinct(KEGG_ko))
  #count(distinct(KEGG_ko))

count_functions_KO <- mags_tax_gene_id_eggnog %>%
  select(eggNOG_OGs, KEGG_ko) %>%
  group_by(KEGG_ko) %>%
  summarise(count = n_distinct(eggNOG_OGs))
  #count(distinct(KEGG_ko))


```

## Df of association between eggnogs and species (as parameters)

```{r, message=F,warning=F, fig.height=9, fig.width = 9}

## Df of association between eggnogs and species of interest

name_test_species <- "association_eggnogs_species.csv"

if (!(name_test_species %in% list.files())) {
  ## if the file is not yet saved, launch the computing

## for each species and for each eggnog, do a chi squared test

heat_evaluate_chi_gathered <- heat_evaluate_chi %>%
  rownames_to_column("MAGS") %>%
  inner_join(mags_tax %>%
               select_('assigned_species') %>%
               rownames_to_column("MAGS"),
             by = "MAGS") %>%
  select(-"MAGS") %>%
  gather("eggnog", "presence", -c('assigned_species'))
  
  
      ## eggnog associated with partial species (1 eggnog only present in 1 species for instance)
      ## must be deleted for chi squared test
eggnog_to_filter <- heat_evaluate_chi_gathered %>%
  group_by_("eggnog", "assigned_species") %>%
  count_("presence") %>%
  ungroup()  %>%
  select(-n, -presence) %>%
  group_by(eggnog) %>%
  count_() %>%
  filter( n == length(species) * 2) %>%
  .$eggnog %>%
  unique()


prevalences <- heat_evaluate_chi_gathered %>%
        ungroup() %>%
        mutate_at("presence", as.integer) %>%
        group_by_("assigned_species", "eggnog") %>%
        summarise(prevalence = (sum(presence)/n()))%>%
        ungroup()

  ##
heat_evaluate_chi_gathered <- heat_evaluate_chi_gathered %>%
  filter(eggnog %in% eggnog_to_filter)

  
  ## create a vector to identify the modalities we test with presence


dim <- heat_evaluate_chi_gathered %>%
  ## do one chi square test per eggnog
  group_by(eggnog) %>%
  nest() %>% 
  mutate(
    chi_test = map(data, ~ chisq.test(.[["assigned_species"]], .$presence)),
    tidied = map(chi_test, tidy)
    ) %>% 
  unnest(tidied) %>%
  ## delete useless columns
  select(-data,- chi_test) %>%
  ## which parameter we compared
  ungroup() %>%
  inner_join(mags_tax_gene_id_eggnog %>% dplyr::rename("eggnog" = "eggNOG_OGs") %>% select(eggnog, Description, EC, KEGG_ko, CAZy), by = "eggnog") %>%
  ungroup() %>%
  distinct() %>%
  arrange(p.value)
      

### merge the prevalences of each species in each row

for (one_species in species){
        
  dim <- dim %>%
    inner_join(prevalences %>%
                 filter(assigned_species == c(one_species)) %>% select(-assigned_species), by = c("eggnog")) %>%
          distinct() %>%
          ## change the column name to prev_modality.name
          rename_at("prevalence", ~ paste0("prevalence_", gsub(" ","_", one_species)))
        
      }
  

write.csv2(dim, name_test_species, quote = F, col.names = TRUE, row.names = FALSE)


data_chi_species <- dim

  
  } else{
  
    data_chi_species <- readr::read_csv2(name_test_species)
  
    }

```


## Df of association between eggnogs and parameters of interest

```{r, message=F,warning=F, fig.height=9, fig.width = 9}


parameters <- c("grp_healthy","westernized", "age_category")


for (parameter in parameters){
  
  
  name_test_parameters <- paste0("association_eggnogs_metadata_", parameter, ".csv")
    
  
  if (!(name_test_parameters %in% list.files())) {
    
    
    data_chi_species <- vector("list")
  
    
    cat(paste0("\n... computing ", parameter, " ...\n\n\n"))

    
    for (i in 1:length(species)) {

      
      ## for each species and for each eggnog, do a chi squared test
      one_species <- species[[i]]
    
      cat(paste0(i, ". ",one_species, "\n"))

      heat_evaluate_chi_gathered <- heat_evaluate_chi %>%
        rownames_to_column("MAGS") %>%
        inner_join(mags_tax %>%
                     select_(parameter, 'assigned_species') %>%
                     rownames_to_column("MAGS"),
                   by = "MAGS") %>%
        select(-"MAGS") %>%
        ## filter the species of interest
        filter(assigned_species == one_species) %>%
        gather("eggnog", "presence", -c(parameter, 'assigned_species'))

      

      prevalences <- heat_evaluate_chi_gathered %>%
        ungroup() %>%
        mutate_at("presence", as.integer) %>%
        group_by_(parameter, "assigned_species", "eggnog") %>%
        summarise(prevalence = (sum(presence)/n()))%>%
        ungroup()
      
      ## create a matrix that will get merged after
      
      prevalences_matrix <- prevalences %>%
        select(-assigned_species) %>%
        reshape2::dcast(paste("eggnog", "~", parameter), value.var = "prevalence", fill=0) %>%
        #reshape2::dcast(eggnog~age_category, value.var = "prevalence", fill=0) %>%
        column_to_rownames("eggnog") %>%
        rename_all(~ paste0("prevalence_", .x)) %>%
        mutate(sum_all_prevalence = rowSums(across(where(is.numeric)))) %>%
        rownames_to_column("eggnog")
      
      #eggnog_to_filter, on enlève ceux qui ont 100 ou 0%
      
      eggnog_to_filter <- prevalences %>%
        filter(!prevalence %in% c(0,1)) %>%
        .$eggnog %>%
        unique()
      
      #eggnog_to_filter, enlever les paramètres associés à une seule modalité
      
      #this eggnog has only yes as westernized
      #28IDW@1|root,2Z8G1@2|Bacteria,2I8P7@201174|Actinobacteria,4D07A@85004|Bifidobacteriales
      
      eggnog_one_modality <- heat_evaluate_chi_gathered %>%
        select_("eggnog",parameter) %>%
        distinct() %>%
        select_("eggnog") %>%
        group_by(eggnog) %>%
        count_() %>%
        filter(n==1) %>%
        select(eggnog) %>%
        distinct()
      
      eggnog_to_filter <- eggnog_to_filter[!eggnog_to_filter %in% eggnog_one_modality$eggnog]
      
  ##
      heat_evaluate_chi_gathered <- heat_evaluate_chi_gathered %>%
        filter(eggnog %in% eggnog_to_filter)
  
  ## create a vector to identify the modalities we test with presence

      modalities <- levels(factor(mags_tax[[parameter]]))

      modalities_vec <- modalities[[1]]

      for (modality in modalities[-1]){
        modalities_vec <- paste0(modality,"-",  modalities_vec)
        }


      dim <- heat_evaluate_chi_gathered %>%
        ## do one chi square test per eggnog
        group_by(eggnog, assigned_species) %>%
        nest() %>% 
        mutate(
          chi_test = map(data, ~ chisq.test(.[[parameter]], .$presence)),
          tidied = map(chi_test, tidy)
          ) %>% 
        unnest(tidied) %>%
        mutate(parameter = c(parameter)) %>%
        ## delete useless columns
        select(-data,- chi_test, -parameter) %>%
        ## which parameter we compared
        mutate(variable = c(parameter)) %>%
        ## which modalities have this parameters
        mutate(variable_modalities = modalities_vec) %>%
        ungroup()
      
      if (nrow(dim) > 0) {
        dim <- dim %>%
          mutate(p.value = p.value %>% p.adjust(method="fdr"))
        }
      
      ### we add the 100% and 0% eggnog but without 
      
      eggnog_not_in_chi <- prevalences %>%
        filter(!eggnog %in% eggnog_to_filter) %>%
        select(assigned_species, eggnog) %>%
        mutate(statistic = "-",
               p.value = 1,
               method = "-",
               variable = parameter,
               variable_modalities = modalities_vec
               )
      
      dim_after <- dim %>%
        rbind(eggnog_not_in_chi %>%
                filter(!eggnog %in% dim$eggnog))
      
      ### add the prevalences per modality
      
      dim_after <- dim_after %>%
        inner_join(prevalences_matrix, by = c("eggnog")) %>%
        ## filter the eggnog which have a prevalence =0 everywhere
        filter(sum_all_prevalence != 0) %>%
        select(-sum_all_prevalence) %>%
        distinct()
      
      
      
      if (nrow(dim_after) > 0) {
        ## if the species is linked to at least 1 eggnog
        data_chi_species[[i]] <- dim_after
        ## filter the value which have 0 everywhere prevalence
        
        
        }
    
    }
    
      data_paramaters <- bind_rows(data_chi_species, .id = "column_label") %>%
    select(-column_label) %>%
    ## merge the eggnog biological description for interpretation
    inner_join(mags_tax_gene_id_eggnog %>% dplyr::rename("eggnog" = "eggNOG_OGs") %>% select(eggnog, Description, EC, KEGG_ko, CAZy), by = "eggnog") %>%
    ungroup() %>%
    distinct()
      
      assign(paste0("data_",parameter), data_paramaters)
      
      ## write data
      
      
      write.csv2(data_paramaters, name_test_parameters, quote = F, col.names = TRUE, row.names = FALSE)

  }else{
    
    assign(paste0("data_",parameter), readr::read_csv2(name_test_parameters))

  
    }
  
  }

```

## Compare every species with other species

```{r, message=F,warning=F, fig.height=9, fig.width = 9}


name_test_species <- "association_species_species.csv"

if (!(name_test_species %in% list.files())) {
  
  data_chi_species <- vector("list")
  
  for (i in 1:length(species)){
    
    one_species <- species[[i]]
    
    ## transform mags_tax to have a new column with this species and other species (mutate)
    
    mags_tax_species <- mags_tax %>% mutate(., species =
                                              case_when(
                                                assigned_species == c(one_species) ~ c(one_species),
                                                assigned_species != c(one_species) ~ "Other species"
                                                ))
    
    parameter <- "species"
    
    cat(paste0("\n... computing ", one_species, " ...\n\n\n"))

    heat_evaluate_chi_gathered <- heat_evaluate_chi %>%
      rownames_to_column("MAGS") %>%
      inner_join(mags_tax_species %>%
                   select_(parameter, 'assigned_species') %>%
                   rownames_to_column("MAGS"),
                 by = "MAGS") %>%
      select(-"MAGS") %>%
      ## filter the species of interest
      gather("eggnog", "presence", -c('assigned_species')) %>%
      mutate(., species =
               case_when(
                 assigned_species == c(one_species) ~ c(one_species),
                 assigned_species != c(one_species) ~ "Other species"
                         ))
    
    
      ## eggnog associated with partial variables (associated with only healthy subjects for instance)
      ## must be deleted for chi squated test
      eggnog_to_filter <- heat_evaluate_chi_gathered %>%
        select_("eggnog", parameter, "presence") %>%
        group_by_("eggnog", parameter, "presence") %>%
        count_(parameter) %>%
        ungroup()  %>%
        select(-n, -presence) %>%
        group_by(eggnog) %>%
        count_(parameter) %>%
        select(-parameter) %>%
        ungroup()  %>%
        group_by(eggnog) %>%
        mutate(n = sum(n)) %>%
      ## we need to filter eggnogs which have the maximum of :
      ## 2 (max nb of modalities for parameter presence) * nb of modalities of parameter (2 for healthy-unhealthy)
        filter(n == 2 * length(levels(factor(mags_tax_species[[parameter]])))) %>%
        select(eggnog) %>%
      .$eggnog
      
      ### we will add a modality for each eggnog which have only 1 modality (0 or 1)
      eggnog_add_chi <- heat_evaluate_chi_gathered %>%
        mutate(., have_one_modality =
                 case_when(
                   eggnog %in% eggnog_to_filter ~ "no",
                   !eggnog %in% eggnog_to_filter ~ "yes"
                   )) %>%
        filter(have_one_modality == "yes")
      
      # 28H6R@1|root,2Z7J3@2|Bacteria,2I9RC@201174|Actinobacteria,4CZ7J@85004|Bifidobacteriales

      prevalences <- heat_evaluate_chi_gathered %>%
        select(- assigned_species) %>%
        ungroup() %>%
        mutate_at("presence", as.integer) %>%
        group_by_(parameter, "eggnog") %>%
        summarise(prevalence = (sum(presence)/n()))%>%
        ungroup() %>%
        filter(eggnog %in% eggnog_to_filter) %>%
        unique()
      
      
  ##
      heat_evaluate_chi_gathered <- heat_evaluate_chi_gathered %>%
        filter(eggnog %in% eggnog_to_filter)
  
  ## create a vector to identify the modalities we test with presence

      modalities <- levels(factor(mags_tax_species[[parameter]]))


      dim <- heat_evaluate_chi_gathered %>%
        ungroup() %>%
        ## do one chi square test per eggnog
        group_by(eggnog) %>%
        nest() %>% 
        mutate(
          chi_test = map(data, ~ chisq.test(.[[parameter]], .$presence)),
          tidied = map(chi_test, tidy)
          ) %>% 
        unnest(tidied) %>%
        mutate(parameter = c(parameter)) %>%
        ## delete useless columns
        select(-data,- chi_test, -parameter) %>%
        ## which modalities have this parameters
        mutate(compared_species = one_species)%>%
        ungroup()
      
      
      ### add the prevalences per modality
      
      for (modality in modalities){
        
        modality <-
        
        dim <- dim %>%
          inner_join(prevalences %>%
                  filter(!!sym(parameter) == c((modality))) %>%
                    select(-!!sym(parameter)), by = c( "eggnog")) %>%
          distinct() %>%
          ## change the column name to prev_modality.name
          rename_at("prevalence", ~ paste0("prevalence_", modality))
        
      }
      
      dim <- dim %>%
        rename_at(paste0("prevalence_", one_species), ~ "prevalence_species")
      
      if (nrow(dim) > 0) {
        ## if the species is linked to at least 1 eggnog
        data_chi_species[[i]] <- dim
        
        
        }
    

    }
    
      data_paramaters <- bind_rows(data_chi_species, .id = "column_label") %>%
    select(-column_label) %>%
    ## merge the eggnog biological description for interpretation
    inner_join(mags_tax_gene_id_eggnog %>% dplyr::rename("eggnog" = "eggNOG_OGs") %>% select(eggnog, Description, EC, KEGG_ko, CAZy), by = "eggnog") %>%
    ungroup() %>%
    distinct() %>%
        mutate(p.value = p.value %>% p.adjust(method="fdr")) %>%
        rename_at("prevalence_Other species", ~ "prevalence_Other_species")
      
      assign("data_species_species", data_paramaters)
      
      ## write data
      
      
      write.csv2(data_paramaters, name_test_species, quote = F, col.names = TRUE, row.names = FALSE)

  }else{
    
    assign("data_species_species", readr::read_csv2(name_test_species))

  
    }


```

## Compare every species with other species

```{r, message=F,warning=F, fig.height=9, fig.width = 9}

## get the core genome


core_genome <- data_species_species %>%
  distinct() %>%
  filter(p.value > 0.1 &
           prevalence_species > 0.9 &
           prevalence_Other_species > 0.9) %>%
  select(-c("statistic", "p.value", "method", "prevalence_species","prevalence_Other_species")) %>%
  distinct() %>%
  ungroup() %>%
  group_by(eggnog) %>%
  count() %>%
  filter(n == 5) %>%
  select(-n) %>%
  inner_join(mags_tax_gene_id_eggnog %>% dplyr::rename("eggnog" = "eggNOG_OGs") %>% select(eggnog, Description, EC, KEGG_ko, CAZy), by = "eggnog") %>%
  distinct()
  

  

## get the specific genes to each species

specific_genome <- data_species_species %>%
  distinct() %>%
  filter(p.value < 0.01 &
           prevalence_species > 0.9 &
           prevalence_Other_species < 0.2) %>%
  group_by(eggnog) %>%
  ### allow to filter the eggnog associated with 1 species only
   filter(n()==1) %>%
  distinct()

```


```{r, message=F,warning=F, fig.height=9, fig.width = 9}

threshold = 0

core_genome <- mags_ko_table %>%
  column_to_rownames("genome_name") %>%
  group_by_at("assigned_species") %>%
  ## calculate prevalence of eggnog by species
  summarise_all(funs(mean(.>0))) %>%
  column_to_rownames("assigned_species") %>%
  ## delete eggnog with prevalence < 90% in all species
  select_if(~all(. > threshold)) %>%
  t %>%
  ## merge annotations
  data.frame(.) %>%
  rownames_to_column("eggnog") %>%
  select(eggnog)
  
  
annotations <- mags_tax_gene_id_eggnog %>%
  filter(completeness > 80) %>%
  select(Description,eggNOG_OGs, EC, KEGG_ko, CAZy) %>%
  distinct() %>%
  merge(core_genome, by.x = "eggNOG_OGs", by.y = "eggnog")


name_core_genome <- paste0("core_genome_annotation_", threshold,".tsv")

if (!name_core_genome %in% list.files()){
  
  write_tsv(annotations, name_core_genome)
  
  core_genome <- annotations
  
}else{
  core_genome <- fread(name_core_genome)
}


```

## Bifidobacterium shunts

```{r, message=F,warning=F, fig.height=9, fig.width = 9}

bifido_shunt <- fread("pathway-genes.txt") %>%
  select(5)
colnames(bifido_shunt) <- "EC"


core_genome_shunt <- core_genome %>%
  merge(bifido_shunt, by = "EC")
print(core_genome_shunt)


whole_genome_shunt <- mags_ko_table %>%
  select(-genome_name) %>%
  t() %>%
  data.frame() %>%
  rownames_to_column("eggnog") %>%
  select(eggnog) %>%
  distinct() %>%
  merge(mags_tax_gene_id_eggnog %>%
          filter(completeness > 80) %>%
          select(eggNOG_OGs, EC) %>%
          distinct(), by.y = "eggNOG_OGs", by.x = "eggnog") %>%
  merge(bifido_shunt, by = "EC")


cat(paste0("we have ", nrow(core_genome_shunt), " bifido shunts in core genome\n"))

cat(paste0("we have ", nrow(whole_genome_shunt), " bifido shunts in whole genome"))

```


## PCA analysis on species prevalences

```{r, message=F,warning=F}

heatmap_prevalence_eggnog_species <- readr::read_csv2("heatmap_prev_species.csv") %>%
  column_to_rownames("X1") %>%
  ## we transpose the table as individuals are eggnogs
  t %>%
  data.frame()

colnames(heatmap_prevalence_eggnog_species) <- gsub("\\.", " ", colnames(heatmap_prevalence_eggnog_species))
colnames(heatmap_prevalence_eggnog_species) <- gsub("Bifidobacterium", "B.", colnames(heatmap_prevalence_eggnog_species))

pca.data <- PCA(heatmap_prevalence_eggnog_species, graph = FALSE, scale.unit = TRUE)

fviz_eig(pca.data, addlabels = TRUE)


fviz_pca_var(pca.data, col.var = "cos2",
             gradient.cols = c("#FFCC00", "#CC9933", "#660033", "#330033"), 
                  repel = TRUE) + ggtitle("PCA on eggnogs - Bifidobacterium species")


```


## PCA analysis on bifidotypes

```{r, message=F,warning=F}

## we use heat to construct the eggnog presence/absence * bifidotypes

heat_bifidotypes <- heat %>%
  rownames_to_column("genome_name") %>%
  merge(mags_tax_gene_id_eggnog %>% select(genome_name, sample_name) %>% distinct(), by = "genome_name") %>%
  merge(fread("enterotype_DMM") %>% select(-V1), by.x = "sample_name", by.y = "sample_id") %>%
  select(-sample_name, -genome_name) %>%
  group_by(bifidotype) %>%
  mutate_all(funs(sum(.>0)/n())) %>%
  distinct() %>%
    ## order dataframe rows
  arrange(bifidotype) %>%
  column_to_rownames("bifidotype") %>%
  t %>%
  data.frame()

colnames(heat_bifidotypes) <- gsub("X","",colnames(heat_bifidotypes))

pca.bifodotypes <- PCA(heat_bifidotypes, graph = FALSE, scale.unit = TRUE)

fviz_eig(pca.bifodotypes, addlabels = TRUE)


fviz_pca_var(pca.bifodotypes, col.var = "cos2",
             gradient.cols = c("#FFCC00", "#CC9933", "#660033", "#330033"), 
                  repel = TRUE) + ggtitle("PCA on eggnogs - bifidotypes")
  

```


## PCA analysis on species-bifidotypes

```{r, message=F,warning=F}

## we use heat to construct the eggnog presence/absence * bifidotypes

heat_bifidotypes <- heat %>%
  rownames_to_column("genome_name") %>%
  merge(mags_tax_gene_id_eggnog %>% select(genome_name, sample_name) %>% distinct(), by = "genome_name") %>%
  merge(fread("enterotype_DMM") %>% select(-V1), by.x = "sample_name", by.y = "sample_id") %>%
   mutate(., bifidotype =
                 case_when(
                   bifidotype %in% c(1,2,6) ~ "healthy",
                   bifidotype %in% c(3,4,5) ~ "unhealthy"
                   )) %>%
  mutate(species_bifidotype=paste0(assigned_species, ".", bifidotype)) %>%
  select(-sample_name, -genome_name, -bifidotype) %>%
  group_by(species_bifidotype) %>%
  mutate_all(funs(sum(.>0)/n())) %>%
  distinct() %>%
    ## order dataframe rows
  arrange(species_bifidotype) %>%
  column_to_rownames("species_bifidotype") %>%
  t %>%
  data.frame()

colnames(heat_bifidotypes) <- gsub("\\.","|",colnames(heat_bifidotypes))


```


## PCA analysis on species-bifidotypes

```{r, message=F,warning=F}


one_species <- "Bifidobacterium adolescentis"

one_species <- gsub("Bifidobacterium ", "",one_species)

pca.bifidotypes <- PCA(heat_bifidotypes %>%
                         select(contains(c(one_species))), graph = FALSE, scale.unit = TRUE)

fviz_eig(pca.bifidotypes, addlabels = TRUE)


fviz_pca_var(pca.bifidotypes, col.var = "cos2",
             gradient.cols = c("#FFCC00", "#CC9933", "#660033", "#330033"), 
             repel = TRUE) + ggtitle("PCA on eggnogs - bifidotypes")

### PCA based on all species
  
pca.bifidotypes <- PCA(heat_bifidotypes, graph = FALSE, scale.unit = TRUE)

fviz_eig(pca.bifidotypes, addlabels = TRUE)


fviz_pca_var(pca.bifidotypes, col.var = "cos2",
             gradient.cols = c("#FFCC00", "#CC9933", "#660033", "#330033"), 
             repel = TRUE) + ggtitle("PCA on eggnogs - bifidotypes")
  



```


## PCA analysis on species-bifidotypes

```{r, message=F,warning=F}

#### work on simple prevalence

prevalences <- heat_evaluate_chi_gathered %>%
        select(- assigned_species) %>%
        ungroup() %>%
        mutate_at("presence", as.integer) %>%
        group_by_(parameter, "eggnog") %>%
        summarise(prevalence = (sum(presence)/n()))%>%
        ungroup() %>%
        filter(eggnog %in% eggnog_to_filter) %>%
        unique()

```
