---
title: "MAGS_analysis"
author: "ruben"
date: "22/04/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Librairies

```{r,message=F,warning=F}

library(dplyr)
library(readr)
library(pheatmap)
library(data.table)
library(tidyr)


source(file = "functions.R")

```

## DonnÃ©es

```{r,message=F,warning=F}

load("curated_v3_otu_tax.rda")

## MAGS annotations data
mags_tax_file = readr::read_tsv("data-raw/bif_mags/Bifidobacterium_annotation/annotations_url_metadata_opendata.tsv") %>%
  filter(assigned_genus=="Bifidobacterium") %>%
  select(genome_name,study,sample_name,assigned_species,completeness, average_distance)

# MAGS linked to other ID's
mags_tax_gene_id_eggnog <- readr::read_csv2("data-raw/bif_mags/mags_tax_gene_id_eggnog.csv") %>%
   select(-X1)

```


## compute genes (eggnog,kegg,cazy) prevalence by species


```{r}

mags_tax_gene_id_eggnog %>% 
   filter(completeness>0.80) %>%
   select(genome_name, assigned_species) %>%
   unique() %>%
   group_by(assigned_species) %>%
   summarise(n_genome = n()) -> prevalence_genomes_species

mags_tax_gene_id_eggnog %>% 
   filter(completeness>0.80) %>%
   select(genome_name, assigned_species, KEGG_ko) %>%
   filter(KEGG_ko != "-") %>%
   unique() %>%
   group_by(KEGG_ko,assigned_species) %>%
   summarise(n_genes=n()) %>%
   merge(prevalence_genomes_species, by="assigned_species") %>%
   mutate(p=n_genes/n_genome)   -> prevalence_ko_species


mags_tax_gene_id_eggnog %>% 
   filter(completeness>0.80) %>%
   select(genome_name, assigned_species, CAZy) %>%
   filter(CAZy != "-") %>%
   unique() %>%
   group_by(CAZy,assigned_species) %>%
   summarise(n_genes=n()) %>%
   merge(prevalence_genomes_species, by="assigned_species") %>%
   mutate(p=n_genes/n_genome)   -> prevalence_cazy_species
   


mags_tax_gene_id_eggnog %>% 
   filter(completeness>0.80) %>%
   select(genome_name, assigned_species, seed_ortholog) %>%
   filter(seed_ortholog != "-") %>%
   unique() %>%
   group_by(seed_ortholog,assigned_species) %>%
   summarise(n_genes=n()) %>%
   merge(prevalence_genomes_species, by="assigned_species") %>%
   mutate(p=n_genes/n_genome)   -> prevalence_eggnog_species



```

### prevalence viz


```{r}

prevalence_ko_species %>%
   filter(assigned_species != "N/A") %>%
   group_by(KEGG_ko) %>%
   mutate(max=max(p), min=min(p)) %>%
   ungroup() %>%
   #filter(max>0.05, min < 0.20 ) %>%
    reshape2::dcast(assigned_species~KEGG_ko, value.var = "p", fill=0) %>%
   tibble::column_to_rownames("assigned_species") %>%
   as.matrix() %>%
   pheatmap::pheatmap(show_colnames = FALSE)


prevalence_ko_species %>%
   filter(assigned_species != "N/A") %>%
   group_by(KEGG_ko) %>%
   mutate(max=max(p), min=min(p)) %>%
   ungroup() %>%
   filter(max>0.05, min < 0.20 ) %>%
    reshape2::dcast(assigned_species~KEGG_ko, value.var = "p", fill=0) %>%
   tibble::column_to_rownames("assigned_species") %>%
   as.matrix() %>%
   pheatmap::pheatmap(show_colnames = FALSE)


```




## Choose the functionnality (Eggnog, Cazy, etc)

```{r,message=F,warning=F}

which_function <- readline(prompt=paste("Which functions you want to plot ? : ", "Kegg-KO - ko", "Cazy - ca","", sep="\n"))

while(!which_function %in% c("ko", "ca")){
  which_function <- readline(prompt=paste("Error, please redo : ", "Kegg-KO - ko", "Cazy - ca","", sep="\n"))
}

if(which_function == "ko"){
  ## we exclude predict data
  
  functionality <- "KEGG_ko"
  
} else{
  ## we keep only predict data
  
  functionality <- "CAZy"
  
}


```

## Prepare plot

```{r,message=F,warning=F}


mags_ko_table <- mags_tax_gene_id_eggnog %>%
   filter(completeness >= 90) %>%
   filter((!!sym(functionality)) != "-") %>%
   group_by_at(functionality) %>%
  mutate(ko_n=n()) %>%
   filter(ko_n>10) %>%
   ungroup() %>%
   group_by_at(vars('genome_name',functionality)) %>%
   summarise(n=n()) %>%
   mutate(n=ifelse(n>1,2,n)) %>%
   reshape2::dcast(paste0('genome_name', '~', functionality), value.var = "n", fill=0)



heat <- mags_ko_table %>%
  tibble::column_to_rownames("genome_name") %>%
  as.matrix()

```

## filter heatmap/add metadata

```{r,message=F,warning=F}

sample <- sampleMetadata

## if one ppl has several samples, keep the sample with the highest nb of reads
sample <- unique(setDT(sampleMetadata)[order(subject_id, -number_reads)], by = "subject_id")

metadata <- sample %>%
   ## transform disease and westernized to good format
   transform_disease(.) %>%
   transform_westernized(.) %>%
   select(sample_id, body_site, disease, age_category, gender, country, westernized, BMI, antibiotics_current_use)


## merge with mags_tax

mags_tax <- mags_tax_file %>%
   merge(metadata, by.x = "sample_name", by.y = "sample_id") %>%
   tibble::column_to_rownames("genome_name") %>%
   filter(assigned_species != "N/A") %>%
   filter(grepl("adolescentis|animalis|longum|bifidum|dentium|pseudocatenulatum|catenulatum", assigned_species))

```

## Filter heat

```{r,message=F,warning=F}

## we filter heat lines which are in mags_tax rownames

heat <- heat %>%
   data.frame(.) %>%
   filter(row.names(.) %in% rownames(mags_tax))


# the species
species <- levels(factor(mags_tax$assigned_species))

```

## Delete the species not assigned to MAGS

```{r,message=F,warning=F, fig.height=9, fig.width = 9}

for (one_species in species){
   
   mags_tax_filtered <- mags_tax %>% filter(assigned_species == one_species)
   
   heat_filtered <- heat %>% filter(row.names(.) %in% rownames(mags_tax_filtered))
   
   
   if(nrow(heat_filtered) <10){
      
      species <- species[!species %in% one_species]
      
      }
      
   }

mags_tax <- mags_tax %>% filter(assigned_species %in% species)

```

## Plotting cazy enzymes

```{r,message=F,warning=F, fig.height=9, fig.width = 9}

library("remotes")
library("RColorBrewer")

palette <- brewer.pal(n = 12, name = 'Paired')


## number of species to respresent
nb_species <- length(species)

annotation_colors = list(
  assigned_species = palette[1:nb_species])

## assign names to the list of colors

names(annotation_colors$assigned_species) <- levels(factor(mags_tax$assigned_species))

```


## Plotting one MAGS per species

```{r,message=F,warning=F, fig.height=9, fig.width = 9}

## KO annotations : plot a mags per species



for (one_species in species){
   
   mags_tax_filtered <- mags_tax %>% filter(assigned_species == one_species)
   
   pheat <- pheatmap(heat %>% filter(row.names(.) %in% rownames(mags_tax_filtered)),
                     cutree_rows = 9,
                     show_rownames = FALSE,
                     show_colnames = FALSE,
                     annotation_row = mags_tax_filtered %>%
                        select(disease, westernized, age_category),
                     annotation_colors = annotation_colors,
                     main= paste0(one_species, " MAGS | ", functionality, " annotations")
                     )

   pheat
}

```
